### PR TITLE
fix: widen decimal arithmetic to Decimal256

### LINKED ISSUE #28584: [Bug]: DECIMAL arithmetic does not widen results beyond Decimal128
### Is there an existing issue for the same bug?

- [x] I have checked the existing issues, including open and closed issues.

### Branch Name

main

### Commit ID

fc621e3616d229c7a29d0c80e73ed8eb1997459c

### Other Environment Information

- MatrixOne: locally built from the commit above (`8.0.30-MatrixOne-v1.3.0`)
- MySQL baseline: 8.0.46
- Verification date: 2026-09-10

### Actual Behavior

Arithmetic on valid `DECIMAL` operands remains in Decimal128 whenever the input type is `DECIMAL(38, s)`, even when the expression result needs more than 38 digits and is representable by MatrixOne's Decimal256 type.

For example, MatrixOne reports `Decimal128 Add overflow` for the sum of two valid `DECIMAL(38,0)` values and `Decimal128 Mul overflow` for the square of a valid `DECIMAL(20,0)` value:

```text
99999999999999999999999999999999999999
+ 99999999999999999999999999999999999999

ERROR 1690: Decimal128 Add overflow

99999999999999999999 * 99999999999999999999

ERROR 1690: Decimal128 Mul overflow
```

This is not a storage-domain overflow of either operand. Explicitly promoting the same operands to `DECIMAL(65,0)` returns the exact 39- and 40-digit results. MySQL 8.0.46 also returns both exact results without an explicit cast.

The behavior propagates into predicates, grouping, ordering, and CTAS expressions because evaluation fails before those operators can consume the value.

Boundary controls are consistent: `DECIMAL(19,0) * DECIMAL(19,0)` and `DECIMAL(20,0) * 2` work because their results fit in Decimal128.

### Expected Behavior

When the declared operand precision implies a result wider than Decimal128 but within MatrixOne's supported Decimal256 range, arithmetic type inference should promote the expression to Decimal256 and return the exact result.

### Steps to Reproduce

```sql
DROP DATABASE IF EXISTS decimal_arithmetic_widening;
CREATE DATABASE decimal_arithmetic_widening;
USE decimal_arithmetic_widening;

CREATE TABLE t (
  id INT PRIMARY KEY,
  d38 DECIMAL(38,0),
  d38s DECIMAL(38,18),
  d20 DECIMAL(20,0),
  d19 DECIMAL(19,0)
);

INSERT INTO t VALUES
  (1,
   99999999999999999999999999999999999999,
   99999999999999999999.999999999999999999,
   99999999999999999999,
   9999999999999999999),
  (2,1,0.000000000000000001,2,2);

-- Decimal128 Add overflow; exact result needs 39 digits.
SELECT d38 + d38 FROM t WHERE id = 1;

-- The scaled DECIMAL(38,18) form has the same problem.
SELECT d38s + d38s FROM t WHERE id = 1;

-- Decimal128 Mul overflow; exact result needs 40 digits.
SELECT d20 * d20 FROM t WHERE id = 1;

-- Downstream relational paths inherit the expression failure.
SELECT id FROM t WHERE d38 + d38 > d38 ORDER BY id;
SELECT d38 + d38 AS k, COUNT(*) FROM t GROUP BY k ORDER BY k;
SELECT id, d38 + d38 AS k FROM t ORDER BY k;

-- Controls that fit in Decimal128.
SELECT d19 * d19 FROM t WHERE id = 1;
SELECT d20 * 2 FROM t WHERE id = 1;

-- Decimal256 proves that MatrixOne can represent both exact results.
SELECT
  CAST(d38 AS DECIMAL(65,0)) + CAST(d38 AS DECIMAL(65,0)),
  CAST(d20 AS DECIMAL(65,0)) * CAST(d20 AS DECIMAL(65,0))
FROM t WHERE id = 1;
```

### Additional information

The arithmetic overload return-type functions in `pkg/sql/plan/function/list_operator.go` return `DECIMAL128(38, scale)` for Decimal128 `+`, `-`, and `*` operands. The execution functions in `pkg/sql/plan/function/arithmetic.go` consequently select the Decimal128 kernels. Decimal256 kernels already exist and work when an operand is explicitly typed as Decimal256.

The complete matrix above was reproduced three times on the same latest-main build and compared with MySQL 8.0.46.

### PR DESCRIPTION
**Pull request by daviszhen**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

Fixes #28584

## What this PR does / why we need it:

- Promotes Decimal128 addition, subtraction, and multiplication to Decimal256 when the declared result requires more than 38 digits, including mixed Decimal64/Decimal128, BIT, and signed/unsigned integer operands.
- Preserves explicit DECIMAL cast domains and canonicalizes genuinely weak decimal literals, including scientific notation, before overload selection.
- Enforces the public 65-digit DECIMAL result limit at runtime so wider Decimal256 coefficients cannot escape under DECIMAL(65) metadata.
- Prevents comparison optimization from narrowing a wide Decimal256 expression back to Decimal128 and keeps affine SUM rewrites compatible with widened arithmetic.
- Adds 65-digit success and 66-digit overflow boundaries, comparison and CTAS coverage, mixed-domain and BIT/DECIMAL resolver coverage, and updates affected decimal BVT expectations.

## Testing

- `go test ./pkg/container/types ./pkg/sql/plan/function ./pkg/sql/plan ./pkg/sql/compile ./pkg/sql/colexec/external -count=1`
- `make build`

### CODE REVIEW

File: pkg/sql/plan/function/list_operator.go [hunk 1]
```diff
@@ -84,6 +84,63 @@ func isDatetimeTimestampComparison(left, right types.Type) bool {
 		left.Oid == types.T_timestamp && right.Oid == types.T_datetime
 }
 
+func decimal128ArithmeticNeedsWidening(operator string, inputs []types.Type) bool {
+	if len(inputs) != 2 || inputs[0].Oid != types.T_decimal128 || inputs[1].Oid != types.T_decimal128 {
```
Comment (aptend) [inline pkg/sql/plan/function/list_operator.go]: [P1] Apply the widening decision after coercion as well as to two already-Decimal128 inputs. With DECIMAL(38,0) + BIGINT, this guard returns false because the right input is integer; fixedTypeCastRule1 then converts both operands to Decimal128, but no widening check runs again, so the operation still uses the Decimal128 kernel. The same hole exists for subtraction and multiplication (and for mixed Decimal64/Decimal128 shapes). A concrete row d=99999999999999999999999999999999999999, i=1 makes d+i need 39 digits and still overflow; d*i with i=2 does too. An exact-head temporary resolver test confirmed that +, -, and * all publish T_decimal128 for DECIMAL(38,0) with BIGINT. Derive precision from the original numeric domains or re-run the precision test on the coerced types before choosing the overload, and add mixed signed/unsigned/integer/decimal boundary cases.

File: pkg/sql/plan/base_binder.go [hunk 2]
```diff
@@ -7175,6 +7183,64 @@ func decimalStringLiteralValue(expr *Expr) (string, bool) {
...
+		switch value := lit.Value.(type) {
+		case *plan.Literal_Decimal64Val:
+			formatted = types.Decimal64(value.Decimal64Val.A).Format(inputs[i].Scale)
+		case *plan.Literal_Decimal128Val:
+			formatted = types.Decimal128{
+				B0_63:   uint64(value.Decimal128Val.A),
+				B64_127: uint64(value.Decimal128Val.B),
+			}.Format(inputs[i].Scale)
+		case *plan.Literal_Sval:
+			formatted = value.Sval
```
Comment (aptend) [inline pkg/sql/plan/base_binder.go]: [P1] Preserve an explicit DECIMAL cast's declared precision, or canonicalize its value at the declared scale before refining it. This loop unwraps every cast, and the Sval branch counts raw characters: CAST('0.5' AS DECIMAL(38,18)) becomes lookup type DECIMAL(2,18). In d38 + that expression, widening then inserts a Decimal128-to-Decimal256 cast with target DECIMAL256(2,18); decimal128ToDecimal256 reparses 0.500000000000000000 against width 2 and returns out-of-range before addition. A temporary exact-head binder test confirmed the published right argument has width=2, scale=18. Scientific notation is also undercounted because exponent digits are counted rather than applied. Do not unwrap syntax-explicit casts; for genuinely weak literals, compute a canonical precision that is at least the scale (including the leading fractional zero) and accounts for exponent/rounding. Add this SQL shape and exponent literals to the tests.

File: pkg/sql/plan/function/list_operator.go [hunk 3]
```diff
@@ -84,6 +84,63 @@ func isDatetimeTimestampComparison(left, right types.Type) bool {
...
+		scale = scale1
+	}
+	if scale2 > scale {
+		scale = scale2
+	}
+	if scale1+scale2 < scale {
+		scale = scale1 + scale2
+	}
+	if parameters[0].Oid == types.T_decimal256 || parameters[1].Oid == types.T_decimal256 {
+		return types.New(types.T_decimal256, min(parameters[0].Width+parameters[1].Width, int32(65)), scale)
```
Comment (aptend) [inline pkg/sql/plan/function/list_operator.go:167]: [P1] Enforce the 65-digit result domain instead of only clamping its metadata. DECIMAL(38,0) * DECIMAL(38,0) selects Decimal256 and this line publishes DECIMAL256(65,0), but d256Mul validates only 256-bit arithmetic, whose internal coefficient can hold 76 decimal digits. For two values consisting of 38 nines, a temporary exact-head kernel test returned a 76-digit product with no error while resolved.retType.Width was 65. That leaves downstream consumers/CTAS with a value outside its declared type rather than the required out-of-range result. Carry the published result width into execution (or otherwise validate the coefficient) and add 65-digit-success/66-digit-overflow boundaries, including CTAS and comparison consumers.

**Review by aptend — CHANGES_REQUESTED** [review]
Deep review of exact head ac9cee1624b513c7f9351db2fab8bdaa9253588d (merge-base 9a07655c54efdd1b5d24f801f701602b7231d013). I read the linked issue, the only issue comment, and checked all review/inline/thread history (there are no prior reviews or threads), then traced the complete six-file diff through overload selection, binder-inserted casts, Decimal128/256 kernels, comparison narrowing, affine SUM rewriting, and downstream published result types. The same-type 39/40-digit cases are correctly widened and the comparison/affine adaptations are directionally sound, but three uncovered boundaries remain blocking: mixed numeric operands still stay on Decimal128, explicit casted string literals can generate an invalid width<scale implicit cast, and products beyond the public 65-digit precision can escape inside a value tagged DECIMAL(65). Native dependencies were built. Full exact-head ./pkg/sql/plan and ./pkg/sql/plan/function tests pass, as do the changed focused tests under -race. Three review-only counterexamples reproduced the findings and were removed; git diff --check is clean.

File: pkg/sql/plan/function/list_operator.go (hunk 1 above)
Comment (daviszhen) [reply pkg/sql/plan/function/list_operator.go]: Fixed in 657a8f5d12. The arithmetic resolver now runs normal coercion first, then derives the widening decision from the original decimal/integer domains. This covers DECIMAL(38) with signed/unsigned integers and mixed Decimal64/Decimal128 operands while retaining Decimal128 for fitting cases such as DECIMAL(20) * literal 2. I also added the missing INT/UINT -> Decimal256 cast-table symmetry found by the TPC-H Q20 regression.

File: pkg/sql/plan/base_binder.go (hunk 2 above)
Comment (daviszhen) [reply pkg/sql/plan/base_binder.go]: Fixed in 657a8f5d12. Syntax-explicit casts are now arithmetic-domain boundaries and are no longer unwrapped. Genuinely weak string literals use PreparedDecimalRuntimeDomains plus Decimal256 parsing to derive canonical precision, including exponent forms, with width kept at least scale. Added binder tests for CAST('0.5' AS DECIMAL(38,18)), 1e2, 1e-2, and leading/trailing zeroes.

File: pkg/sql/plan/function/list_operator.go (hunk 3 above)
Comment (daviszhen) [reply pkg/sql/plan/function/list_operator.go:167]: Fixed in 657a8f5d12. Decimal256 +, -, and * now validate each non-NULL result coefficient against the published DECIMAL(p,s) precision before it can reach storage or another consumer. Tests cover a 65-digit successful product, a 66-digit overflow, comparison type preservation, and CTAS DECIMAL(65,0) metadata. The affected decimal BVT expectations were also updated from the old Decimal128 overflow behavior; the overflow case was verified against an isolated Docker MatrixOne and returns SQLSTATE 22003.

**Comment by daviszhen** [comment]
已修复本轮 CI 的两个实际代码问题，提交 e75d294e7c：

1. Compose PROXY BVT
- DECIMAL(38,0) * 2 现在按精确的 39 位 Decimal256 结果更新测试预期，不再期待旧的 Decimal128 overflow。
- 仅对乘法使用整数 literal 的实际精度做 overload lookup，并确保 lookup 后的类型 refinement 会生成真实 cast。
- 加减保持原有 Decimal128 coercion，修复 RIGHT JOIN 条件 l.v <= r.v + 8 丢失 (2,12) 的问题。

2. UT Coverage producer
- 精确 DECIMAL widening 会使 13 个 canonical TPC-H plan 包含 Decimal256；Substrait DECIMAL 最大 precision 是 38，因此测试明确验证这些 plan 安全返回 not eligible，其余 plan 仍必须成功导出。
- Coverage 汇总作业失败是该 producer 失败的级联结果。

普通 Ubuntu UT 的另一个失败是作业达到 70 分钟上限时 TestV406UpgradeRecoversLegacyTinyText 仍在运行，被 timeout 124 终止，没有出现与本 PR 相关的断言失败；新提交已重新触发。

本地验证：
- go test ./pkg/sql/plan ./pkg/sql/plan/function ./pkg/sql/plan/substrait -count=1：通过
- make build：通过
- 隔离 Docker Compose + 最新 mo-tester 334d3e1 执行 dtype/decimal.test：688/688，通过；所有测试容器 restart count 为 0。

**Comment by daviszhen** [comment]
Updated to current upstream/main (cd8e8d9134) in 109931461e.\n\nThe latest ordinary UT failure was the overall UT deadline (exit 124), not a Decimal assertion. The coverage producer failed earlier while setup-java was downloading JDK 8 and hit the action timeout; downstream Coverage/CI Required then failed by cascade. Both compose BVT jobs passed in that run. The branch now includes main's bounded UT overlap and timeout scheduling update (#28643).\n\nVerification:\n- make build\n- go test ./pkg/sql/plan/function ./pkg/sql/plan -count=1\n- golangci-lint v2.6.2 on affected packages: 0 issues\n\nNew CI: https://github.com/matrixorigin/matrixone/actions/runs/34557994277

**Comment by daviszhen** [comment]
Resolved the latest failing checks at head `d8342a4199` and merged current upstream `main` through `b349fea56f`.

The diagnostics show no Decimal regression: coverage and both Compose/Standalone BVT jobs passed. The only failed producer was ordinary Ubuntu UT:

- `pkg/tests/issues/TestWWConflict` created its 10-second context before shared-cluster database/table setup;
- under race/CI load, setup consumed that deadline and txn1 returned `context deadline exceeded`;
- txn2 then waited forever on `txn2StartedC`, so the package reached its 40-minute timeout and the remaining embedded stage was terminated by the overall job deadline.

Fix:

- start a 30-second context only after fixture setup, so it bounds the concurrent transaction workflow itself;
- make waits for `txn2StartedC` and `txn2CommittedC` context-aware, preventing a failed transaction from leaving its peer blocked forever.

Local validation passed:

- `go test -tags matrixone_test -race ./pkg/tests/issues -run '^TestWWConflict$' -count=1 -timeout 10m` (15.989s);
- `go test ./pkg/sql/plan ./pkg/sql/plan/function ./pkg/sql/plan/substrait -count=1`;
- `make build`;
- `git diff --check`.

A fresh CI run has been triggered by the push.

**Review by XuPeng-SH — APPROVED** [review]
Reviewed the full current change at d8342a41999e3b17df8bde55c3cfe23314ff942f against main b349fea56f1c16e3c947ac730cc114fca3c3f241, linked issue #28584, and all earlier review findings/replies. No remaining concrete blocker found.

All three previous findings are closed: widening is derived from original decimal/signed/unsigned domains after coercion; syntax-explicit CAST remains a domain boundary and weak literals are canonicalized at the declared scale; Decimal256 +, -, and * now validate non-NULL result coefficients against the published precision. The literal lookup refinement is materialized as an actual argument cast, avoiding a dispatch-only type change. Comparison narrowing now checks integral capacity as well as scale.

Traced binder/result metadata, existing Decimal256 kernels, constant/vector/NULL/selection-mask handling, casts and CTAS consumers, and the affine SUM proof/cast-back path. Existing physical operator IDs and Decimal256 representation are reused, rather than changing a legacy overload's physical argument/result contract. Affine restoration remains behind its exact integer-range/cardinality proof. Wider plans deliberately decline Substrait export instead of silently truncating to its 38-digit domain; this is a real offload/performance tradeoff explicitly recorded by the changed tests. Narrow fitting paths remain available; widened execution pays bounded 32-byte vector/cast work and one precision-validation pass.

The accompanying TestWWConflict adjustment starts its budget after setup, bounds both channel waits, retains transaction error assertions, joins both workers, and preserves deferred hook restoration. It does not retry away an assertion failure or remove the conflict scenario.

Validation: source/diff and consumer reasoning; reviewed mixed-domain resolver, explicit-cast/exponent, 65/66-digit runtime, comparison/CTAS, affine and BVT oracles. Reused the author's current-head reported plan/function/substrait package, targeted race TestWWConflict and build results, plus earlier relevant decimal BVT evidence. No fresh native suite, SQL deployment, mixed-version run or benchmark was executed by this reviewer; CI was not awaited. The PR body's older failed-build fragments are not counted as passing evidence. Historical executors still retain their own pre-fix precision-error behavior until upgraded; no new wire/disk representation is introduced.

**Comment by daviszhen** [comment]
已确认 aptend 提出的三个 Decimal256 精度边界均已落实：mixed numeric coercion 后 widening、显式 DECIMAL cast/弱字面量 domain 处理、以及 Decimal256 结果 65 位精度校验；对应 threads 均已解决，XuPeng-SH 已在上一 head 批准。

本次已同步当前 upstream/main（23b20ea909），新 head 为 a2fb9b1d7e。

验证：
- go test ./pkg/sql/plan ./pkg/sql/plan/function ./pkg/sql/plan/substrait -count=1：通过
- make build：通过
- git diff --check：通过

请复审当前 head。

**Comment by daviszhen** [comment]
Synced the latest main at head 62c370f294 after main advanced during review. The Decimal256 precision constraints remain intact. Revalidated go test ./pkg/sql/plan ./pkg/sql/plan/function ./pkg/sql/plan/substrait -count=1 and make build; all pass.

**Comment by daviszhen** [comment]
Synced with the latest upstream main (6602ca0ffa) and resolved the arithmetic cast-rule conflict.\n\nConflict resolution:\n- preserve main's arithmeticTypeCastRule1 behavior for BIT unsigned range\n- then apply widenedDecimalArithmeticInputs for DECIMAL +, -, and *\n\nLocal verification:\n- go test ./pkg/sql/plan ./pkg/sql/plan/function ./pkg/sql/plan/substrait -count=1\n- make build\n\nAll passed. The branch is no longer behind main and GitHub reports it as mergeable. New head: 3383f8c58845511f09dc7bbd82a83e73f6dff52e.

**Comment by daviszhen** [comment]
Fixed the failing PESSIMISTIC Compose BVT check in `2eebaf00ab`.

Root cause:
- The failure was in `test/distributed/cases/dtype/bit.sql`.
- The golden result still expected `Decimal128 Mul overflow` for `(2^64 - 1)^2`.
- This PR intentionally widens the expression to Decimal256, so the actual exact value `340282366920938463426481119284349108225` is the correct new behavior.

Changes:
- Updated the BIT BVT SQL/result to assert the widened Decimal256 result.
- Synchronized the branch with the latest `main`.

Local verification:
- `go test ./pkg/sql/plan ./pkg/sql/plan/function ./pkg/sql/plan/substrait -count=1` passed.
- `make build` passed.
- `git diff --check` passed.

A new CI run has been triggered for head `2eebaf00ab5ef2aa720f68fc278e847af37f58e7`. Docker is not installed on this local machine, so the exact Compose BVT will be verified by CI.

**Comment by daviszhen** [comment]
Follow-up fix pushed in `bea96bede2`.

The PROXY Compose `NumberFormatException` was caused by the explanatory SQL comments added between two result blocks. mo-tester appended those comments to the preceding numeric cell, then attempted to parse this combined text:

`17878103347812890625 + comment text`

The Decimal256 value itself was not the parse failure. I removed the inter-query comments while retaining the exact widened result assertion.

The other Ubuntu UT failure was `TestISCPExecutor1`: its watermark stayed at `0-0` during the CI wait. This is unrelated to the decimal code path, and the focused race-detector test passes locally:

`go test -tags matrixone_test -race ./pkg/vm/engine/test -run ^TestISCPExecutor1$ -count=1 -timeout 10m`

Also verified after synchronizing latest `main`:
- decimal/plan test packages pass
- `make build` passes
- `git diff --check` passes

New CI run: https://github.com/matrixorigin/matrixone/actions/runs/34674424886

**Comment by daviszhen** [comment]
Resolved the current failing checks by rerunning the failed producer job at the same head `308db3cc5d`; no source change was appropriate.

Root cause:
- PESSIMISTIC Compose BVT attempt 1 failed in `Set up Go And Java` with `read ECONNRESET`, before the shared MatrixOne build was downloaded or any container/BVT started.
- Its coverage artifact was therefore an incomplete 154-byte marker.
- The downstream Coverage job correctly refused to calculate a verdict because a required producer had not published a complete profile.
- The same run's PROXY BVT, Ubuntu UT, UT Coverage, and SCA all passed, so there was no Decimal256 regression to patch.

Job-level rerun results (attempt 2):
- `Set up Go And Java`: passed;
- PESSIMISTIC Compose BVT: passed;
- generated coverage profile: 3,241,634 bytes;
- Coverage merge: passed;
- CI Required: passed.

Run: https://github.com/matrixorigin/matrixone/actions/runs/34763057015

File: pkg/sql/plan/function/list_operator.go [hunk 4]
```diff
@@ -84,6 +84,91 @@ func isDatetimeTimestampComparison(left, right types.Type) bool {
...
+func decimalArithmeticDomain(input types.Type) (width, scale int32, ok bool) {
+	if input.Oid.IsDecimal() {
+		width = input.Width
+		if width <= 0 {
+			width = input.Oid.ToType().Width
+		}
+		scale = max(input.Scale, int32(0))
+		return width, scale, true
+	}
+	if input.IsIntOrUint() {
```
Comment (aptend) [inline pkg/sql/plan/function/list_operator.go:96]: BIT is excluded here because Type.IsIntOrUint() is false for T_bit, even though the fixed coercion table maps BIT with DECIMAL to Decimal128 and BIT(64) carries the full uint64 domain (20 decimal digits). Reproducer: CREATE TABLE t(b BIT(64), d DECIMAL(38,0)); INSERT INTO t VALUES(18446744073709551615, 99999999999999999999999999999999999999); SELECT b + d, b * d FROM t;. At this head both expressions resolve to Decimal128(38); the exact results need 39 and 58 digits respectively, so execution overflows instead of taking the new Decimal256 path. A temporary resolver test for both + and * failed with expected Decimal256 / actual Decimal128. Please include BIT as a 20-digit exact arithmetic domain and cover both operand orders.

**Review by aptend — CHANGES_REQUESTED** [review]
Re-reviewed the complete diff, all review history and resolved threads, plus the delta since my prior review at ac9cee1624b513c7f9351db2fab8bdaa9253588d, against exact head 308db3cc5d93480beba3182b6d984f873b58b66a. The three previously reported Decimal/integer, explicit-cast/literal, and published-precision issues are closed. One exact numeric coercion domain is still omitted and reproduces the same overflow this PR is intended to eliminate. Exact-head CI is green, but it does not cover this direct BIT/DECIMAL pair.

File: pkg/sql/plan/function/list_operator.go (hunk 4 above)
Comment (daviszhen) [reply pkg/sql/plan/function/list_operator.go:96]: Addressed in 76dc050889. BIT is now modeled as a 20-digit exact arithmetic domain for widening, with resolver regressions for + and * in both operand orders against DECIMAL(38,0). Full pkg/sql/plan/function and pkg/sql/plan tests pass.

**Comment by daviszhen** [comment]
Pushed 83222beafa, merging current upstream/main (bb358ea76a) and restarting CI.

The reported TestIssue28313EmptyRemoteDispatch failure was a 180-second context timeout in the full Ubuntu UT run; the Decimal change does not touch dispatch/shuffle lifecycle code. On the updated head, the exact regression passes three consecutive runs, and the affected plan/function packages pass as well.

**Review by aptend — APPROVED** [review]
Deep re-review completed on exact head `ae7fd9949383d80de87092894865cfa619f80ab3` against exact merge-base `07fdd4ae0f80f287b93fc4b525fd296d5617abc7`. I read the complete review and inline-comment history, author replies, issue discussion, and resolved/unresolved thread state; compared the delta after my prior reviewed head `308db3cc5d93480beba3182b6d984f873b58b66a`; and rechecked the complete arithmetic resolver, binder casts, Decimal256 kernels/precision checks, comparison and affine-SUM consumers, Substrait boundary, tests, and BVT expectations. My remaining BIT/DECIMAL blocker is closed in code: BIT is now treated as a 20-digit exact unsigned domain for the widening decision, both operand orders produce Decimal256 casts, and +/* publish 39/58-digit Decimal256 results. I independently built the plan-level BIT(64)/DECIMAL(38,0) + and * shapes and confirmed their physical arguments/results are Decimal256. The earlier mixed integer/decimal, explicit-cast/weak-literal, and 65-digit runtime-boundary findings remain closed. Validation on this exact head: full `pkg/sql/plan`, `pkg/sql/plan/function`, and `pkg/sql/plan/substrait` suites pass with repository CGO/third-party artifacts; `git diff --check` passes; all required CI checks are green. No remaining blocking issue found.

**Comment by daviszhen** [comment]
Resolved the latest main conflict at 0a490af6cc after preserving the remote branch updates and merging upstream/main through 2ea4ec2e6e.

The list_operator conflict now retains both dispatch layers: main mixed signed/unsigned integer domain handling runs first, then this PR applies widened Decimal256 operand selection for decimal arithmetic. This also integrates main Decimal256 high-scale multiplication coverage without dropping either behavior.

Local validation passed for container/types, plan/function, plan, compile, and external, followed by the complete native-provenance-aware make build.

**Review by aunjgr — CHANGES_REQUESTED** [review]
[P2] Correct the newly accepted wrapped decimal multiplication result.

At test/distributed/cases/dtype/decimal.result:1306-1310, this PR replaces an overflow with three successful values that are mathematically incorrect. The fixture declares col2 DECIMAL(19,0), and its first value is 1235467899687894561. For:

SELECT col2 * -2733892455124775.7851878942123454 FROM decimal19;

the first exact product is:
-3377636369505588272411432821735887.9920303797133694

but the new expected result is:
-12929400140562083461.4885988496007550

The exact coefficient has 50 digits and fits the advertised DECIMAL(51,16), so this is not an allowed overflow or rounding difference. The expected value exactly equals col2 multiplied by the signed low 64 bits of the literal's scaled coefficient: int64(-27338924551247757851878942123454) = -104651850070959550, divided by 10^16. The other two newly accepted rows follow the same truncated coefficient.

Please correct the oracle using independent arbitrary-precision arithmetic and verify the exact-head SQL execution. If execution produces the checked-in values, fix the constant conversion/widened execution path before accepting the result. Add a focused positive/negative wide-fractional-literal case and literal-versus-typed-column control; resolver/type-only and scale-zero multiplication tests cannot catch this.

Reviewed all 16 changed files and traced original-domain widening, explicit-cast/weak-literal handling, physical argument casts, Decimal256 precision checks, comparison/affine consumers and tests. Prior mixed integer/BIT-domain fixes are retained. This finding is independently proven against the checked-in fixture/oracle with exact integer arithmetic; I have not run a fresh SQL reproduction or established the precise runtime truncation site, and am not claiming that a generated expected file proves runtime behavior. Current-head UT and BVT were still running at submission preflight. No other confirmed blocker found.

**Comment by daviszhen** [comment]
Thanks — confirmed on exact head: the checked-in values were the low-64-bit wrapped products.

Root cause was in the Decimal256 fast-path classifiers, not in literal parsing or Decimal128→Decimal256 casting. `d256AllFitInt64` mixed `^` and `|` without grouping the per-limb comparisons. Because those operators have the same precedence and associate left-to-right in Go, the all-ones sign-extension limbs of a wide negative value could cancel the non-zero mismatch. The value was then incorrectly sent through the int64 multiplication tier, which consumed only `B0_63`. `d256AllFitInt32` had the same latent expression and is fixed as well.

Fixed in c9bdffcb6e:
- parenthesized/isolated each upper-limb sign-extension mismatch before OR-reduction;
- corrected all three decimal19 oracle rows using independent arbitrary-precision arithmetic;
- added executing positive/negative wide-fractional-literal SQL regressions;
- added literal-scalar versus typed-column vector controls for both signs;
- added a typed `DECIMAL(32,16)` BVT column control alongside the literal cases.

Exact-head validation:
- `pkg/sql/plan/function`: full package PASS;
- `pkg/sql/plan`: full package PASS;
- focused exact SQL bind/execute now returns `-3377636369505588272411432821735887.9920303797133694` for the reported first row;
- `git diff --check`: PASS.

**Comment by daviszhen** [comment]
Resolved the conflict with current `main` (`091825aac2`).

Conflict decision in `pkg/sql/plan/substrait/tpch_test.go`:
- preserved the Decimal256 arithmetic fallback coverage from this PR;
- included Q6 introduced by Decimal256 SUM on `main`;
- retained exact `type` eligibility assertions for Q6/Q11, while checking the earlier `expression` rejection produced by widened arithmetic in the other affected queries.

New exact head: `0aba14a1b0fff6928e5e4097bfe0e4921eea8251`

Validation passed:
- `make config`
- `go test ./pkg/sql/plan ./pkg/sql/plan/function ./pkg/sql/plan/substrait -count=1` (with the resolved TPCH assertion rerun after refinement)

**Review by aunjgr — APPROVED** [review]
No blocking finding at this revision.

Compared the patch series against my previous reviewed head: all nine non-merge patches are unchanged after the main merge. Rechecked the decimal widening/cast/precision boundaries, comparison non-narrowing, affine-result restoration and changed binder surroundings. The prior full source/test/BVT review remains applicable, and exact-head CI is green.

Existing review and CI evidence reused; no new local test execution or performance claim.
